# Web Scraping Lab

You will find in this notebook some scrapy exercises to practise your scraping skills.

**Tips:**

- Check the response status code for each request to ensure you have obtained the intended contennt.
- Print the response text in each request to understand the kind of info you are getting and its format.
- Check for patterns in the response text to extract the data/info requested in each question.
- Visit each url and take a look at its source through Chrome DevTools. You'll need to identify the html tags, special class names etc. used for the html content you are expected to extract.

- [Requests library](http://docs.python-requests.org/en/master/#the-user-guide) documentation 
- [Beautiful Soup Doc](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [Urllib](https://docs.python.org/3/library/urllib.html#module-urllib)
- [re lib](https://docs.python.org/3/library/re.html)
- [lxml lib](https://lxml.de/)
- [Scrapy](https://scrapy.org/)
- [List of HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
- [HTML basics](http://www.simplehtmlguide.com/cheatsheet.php)
- [CSS basics](https://www.cssbasics.com/#page_start)

#### Below are the libraries and modules you may need. `requests`,  `BeautifulSoup` and `pandas` are imported for you. If you prefer to use additional libraries feel free to uncomment them.

In [1]:
import requests
from bs4 import BeautifulSoup as bs
import pandas as pd
# from pprint import pprint
from lxml import html
# from lxml.html import fromstring
# import urllib.request
# from urllib.request import urlopen
import random
import re
# import scrapy

#### Download, parse (using BeautifulSoup), and print the content from the Trending Developers page from GitHub:

In [2]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/developers'

In [3]:
#your code

req = requests.get(url).content
req

b'\n\n\n\n\n<!DOCTYPE html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8">\n  <link rel="dns-prefetch" href="https://github.githubassets.com">\n  <link rel="dns-prefetch" href="https://avatars.githubusercontent.com">\n  <link rel="dns-prefetch" href="https://github-cloud.s3.amazonaws.com">\n  <link rel="dns-prefetch" href="https://user-images.githubusercontent.com/">\n\n\n\n  <link crossorigin="anonymous" media="all" integrity="sha512-bvmPnIfSXuQ7bosV8Ys+XNjZ7wqcfxwStr1JGjt4ibr/TK0ycsQZAYi4/4aziAgybJgv2/5+bw4lJF7CRSRs/g==" rel="stylesheet" href="https://github.githubassets.com/assets/frameworks-6ef98f9c87d25ee43b6e8b15f18b3e5c.css" />\n  <link crossorigin="anonymous" media="all" integrity="sha512-ZUf6K+vQqMY+RhVzaRmCy2ePbSZad4TkaGRbd6v5gFt6f9Q/nqjDkBDjQgXNmZw7J9mcYxlsE4fhRw7CTluRow==" rel="stylesheet" href="https://github.githubassets.com/assets/site-6547fa2bebd0a8c63e461573691982cb.css" />\n    <link crossorigin="anonymous" media="all" integrity="sha512-OYEON3KwVcOaAjKMGi0Jtd

#### Display the names of the trending developers retrieved in the previous step.

Your output should be a Python list of developer names. Each name should not contain any html tag.

**Instructions:**

1. Find out the html tag and class names used for the developer names. You can achieve this using Chrome DevTools.

1. Use BeautifulSoup to extract all the html elements that contain the developer names.

1. Use string manipulation techniques to replace whitespaces and linebreaks (i.e. `\n`) in the *text* of each html element. Use a list to store the clean names.

1. Print the list of names.

Your output should look like below:

```
['trimstray (@trimstray)',
 'joewalnes (JoeWalnes)',
 'charlax (Charles-AxelDein)',
 'ForrestKnight (ForrestKnight)',
 'revery-ui (revery-ui)',
 'alibaba (Alibaba)',
 'Microsoft (Microsoft)',
 'github (GitHub)',
 'facebook (Facebook)',
 'boazsegev (Bo)',
 'google (Google)',
 'cloudfetch',
 'sindresorhus (SindreSorhus)',
 'tensorflow',
 'apache (TheApacheSoftwareFoundation)',
 'DevonCrawford (DevonCrawford)',
 'ARMmbed (ArmMbed)',
 'vuejs (vuejs)',
 'fastai (fast.ai)',
 'QiShaoXuan (Qi)',
 'joelparkerhenderson (JoelParkerHenderson)',
 'torvalds (LinusTorvalds)',
 'CyC2018',
 'komeiji-satori (神楽坂覚々)',
 'script-8']
 ```

In [4]:
#your code

soup = bs(req, 'html.parser')
print(soup)


<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="utf-8"/>
<link href="https://github.githubassets.com" rel="dns-prefetch"/>
<link href="https://avatars.githubusercontent.com" rel="dns-prefetch"/>
<link href="https://github-cloud.s3.amazonaws.com" rel="dns-prefetch"/>
<link href="https://user-images.githubusercontent.com/" rel="dns-prefetch"/>
<link crossorigin="anonymous" href="https://github.githubassets.com/assets/frameworks-6ef98f9c87d25ee43b6e8b15f18b3e5c.css" integrity="sha512-bvmPnIfSXuQ7bosV8Ys+XNjZ7wqcfxwStr1JGjt4ibr/TK0ycsQZAYi4/4aziAgybJgv2/5+bw4lJF7CRSRs/g==" media="all" rel="stylesheet">
<link crossorigin="anonymous" href="https://github.githubassets.com/assets/site-6547fa2bebd0a8c63e461573691982cb.css" integrity="sha512-ZUf6K+vQqMY+RhVzaRmCy2ePbSZad4TkaGRbd6v5gFt6f9Q/nqjDkBDjQgXNmZw7J9mcYxlsE4fhRw7CTluRow==" media="all" rel="stylesheet">
<link crossorigin="anonymous" href="https://github.githubassets.com/assets/behaviors-39810e3772b055c39a02328c1a2d09b5.css" integr

In [5]:
devs = soup.findAll('h1', {'class':'h3 lh-condensed'})
devs[0].string

#list_names = [devs[name].string for name in range(len(devs))]
#list_names

In [6]:
names = [dev.get_text() for dev in devs]
names

['\n\n            Miek Gieben\n ',
 '\n\n            Ha Thach\n ',
 '\n\n            Alexey Golub\n ',
 '\n\n            Alon Zakai\n ',
 '\n\n            snipe\n ',
 '\n\n            Mo Gorhom\n ',
 '\n\n            Ben Johnson\n ',
 '\n\n            Francois Zaninotto\n ',
 '\n\n            Han Xiao\n ',
 '\n\n            Anthony Sottile\n ',
 '\n\n            Isaac Hinman\n ',
 '\n\n            Dries Vints\n ',
 '\n\n            Mitchell Hashimoto\n ',
 '\n\n            Bruno Garcia\n ',
 '\n\n            GitHub Teacher\n ',
 '\n\n            Chris Long\n ',
 '\n\n            Sebastián Ramírez\n ',
 '\n\n            Ashleigh Simonelli\n ',
 '\n\n            William Candillon\n ',
 '\n\n            Michael Mok\n ',
 '\n\n            Sébastien Lorber\n ',
 '\n\n            文翼\n ',
 '\n\n            Michele Locati\n ',
 '\n\n            Daniel Gerlag\n ',
 '\n\n            Nelson Osacky\n ']

In [7]:
names = [name.replace('\n','') for name in names]
names

['            Miek Gieben ',
 '            Ha Thach ',
 '            Alexey Golub ',
 '            Alon Zakai ',
 '            snipe ',
 '            Mo Gorhom ',
 '            Ben Johnson ',
 '            Francois Zaninotto ',
 '            Han Xiao ',
 '            Anthony Sottile ',
 '            Isaac Hinman ',
 '            Dries Vints ',
 '            Mitchell Hashimoto ',
 '            Bruno Garcia ',
 '            GitHub Teacher ',
 '            Chris Long ',
 '            Sebastián Ramírez ',
 '            Ashleigh Simonelli ',
 '            William Candillon ',
 '            Michael Mok ',
 '            Sébastien Lorber ',
 '            文翼 ',
 '            Michele Locati ',
 '            Daniel Gerlag ',
 '            Nelson Osacky ']

In [8]:
clean = [clean.replace('\n','') for clean in names]
clean

['            Miek Gieben ',
 '            Ha Thach ',
 '            Alexey Golub ',
 '            Alon Zakai ',
 '            snipe ',
 '            Mo Gorhom ',
 '            Ben Johnson ',
 '            Francois Zaninotto ',
 '            Han Xiao ',
 '            Anthony Sottile ',
 '            Isaac Hinman ',
 '            Dries Vints ',
 '            Mitchell Hashimoto ',
 '            Bruno Garcia ',
 '            GitHub Teacher ',
 '            Chris Long ',
 '            Sebastián Ramírez ',
 '            Ashleigh Simonelli ',
 '            William Candillon ',
 '            Michael Mok ',
 '            Sébastien Lorber ',
 '            文翼 ',
 '            Michele Locati ',
 '            Daniel Gerlag ',
 '            Nelson Osacky ']

In [9]:
list_names = [i.replace('  ', '') for i in clean ]
list_names

['Miek Gieben ',
 'Ha Thach ',
 'Alexey Golub ',
 'Alon Zakai ',
 'snipe ',
 'Mo Gorhom ',
 'Ben Johnson ',
 'Francois Zaninotto ',
 'Han Xiao ',
 'Anthony Sottile ',
 'Isaac Hinman ',
 'Dries Vints ',
 'Mitchell Hashimoto ',
 'Bruno Garcia ',
 'GitHub Teacher ',
 'Chris Long ',
 'Sebastián Ramírez ',
 'Ashleigh Simonelli ',
 'William Candillon ',
 'Michael Mok ',
 'Sébastien Lorber ',
 '文翼 ',
 'Michele Locati ',
 'Daniel Gerlag ',
 'Nelson Osacky ']

#### Display the trending Python repositories in GitHub

The steps to solve this problem is similar to the previous one except that you need to find out the repository names instead of developer names.

In [10]:
# This is the url you will scrape in this exercise
url2 = 'https://github.com/trending/python?since=daily'

In [11]:
#your code
html2 = requests.get(url2).content
soup = bs(html2, 'lxml')

repo = soup.findAll('h1', {'class':'h3 lh-condensed'})
list_repo = [i.text for i in repo]
list_repo


['\n\n\n\n        yuchenlin /\n\n      rebiber\n ',
 '\n\n\n\n        qutebrowser /\n\n      qutebrowser\n ',
 '\n\n\n\n        sdushantha /\n\n      wifi-password\n ',
 '\n\n\n\n        PaddlePaddle /\n\n      PaddleOCR\n ',
 '\n\n\n\n        microsoft /\n\n      qlib\n ',
 '\n\n\n\n        mportesi /\n\n      sorting_algo_visualizer\n ',
 '\n\n\n\n        mobvoi /\n\n      wenet\n ',
 '\n\n\n\n        celery /\n\n      celery\n ',
 '\n\n\n\n        PyTorchLightning /\n\n      pytorch-lightning\n ',
 '\n\n\n\n        donnemartin /\n\n      data-science-ipython-notebooks\n ',
 '\n\n\n\n        public-apis /\n\n      public-apis\n ',
 '\n\n\n\n        miloyip /\n\n      game-programmer\n ',
 '\n\n\n\n        facebook /\n\n      prophet\n ',
 '\n\n\n\n        jina-ai /\n\n      jina\n ',
 '\n\n\n\n        streamlit /\n\n      streamlit\n ',
 '\n\n\n\n        apache /\n\n      airflow\n ',
 '\n\n\n\n        JaidedAI /\n\n      EasyOCR\n ',
 '\n\n\n\n        allenai /\n\n      allennlp\n '

In [12]:
clean_repo = [list_repo.replace('\n','') for list_repo in list_repo]
clean_repo

['        yuchenlin /      rebiber ',
 '        qutebrowser /      qutebrowser ',
 '        sdushantha /      wifi-password ',
 '        PaddlePaddle /      PaddleOCR ',
 '        microsoft /      qlib ',
 '        mportesi /      sorting_algo_visualizer ',
 '        mobvoi /      wenet ',
 '        celery /      celery ',
 '        PyTorchLightning /      pytorch-lightning ',
 '        donnemartin /      data-science-ipython-notebooks ',
 '        public-apis /      public-apis ',
 '        miloyip /      game-programmer ',
 '        facebook /      prophet ',
 '        jina-ai /      jina ',
 '        streamlit /      streamlit ',
 '        apache /      airflow ',
 '        JaidedAI /      EasyOCR ',
 '        allenai /      allennlp ',
 '        aws /      aws-cli ',
 '        geerlingguy /      ansible-for-devops ',
 '        spulec /      moto ',
 '        spack /      spack ',
 '        openai /      baselines ',
 '        xonsh /      xonsh ',
 '        owid /      covid-19-dat

In [13]:
repo_ready = [i.replace('    ', '') for i in clean_repo]
repo_ready

['yuchenlin /  rebiber ',
 'qutebrowser /  qutebrowser ',
 'sdushantha /  wifi-password ',
 'PaddlePaddle /  PaddleOCR ',
 'microsoft /  qlib ',
 'mportesi /  sorting_algo_visualizer ',
 'mobvoi /  wenet ',
 'celery /  celery ',
 'PyTorchLightning /  pytorch-lightning ',
 'donnemartin /  data-science-ipython-notebooks ',
 'public-apis /  public-apis ',
 'miloyip /  game-programmer ',
 'facebook /  prophet ',
 'jina-ai /  jina ',
 'streamlit /  streamlit ',
 'apache /  airflow ',
 'JaidedAI /  EasyOCR ',
 'allenai /  allennlp ',
 'aws /  aws-cli ',
 'geerlingguy /  ansible-for-devops ',
 'spulec /  moto ',
 'spack /  spack ',
 'openai /  baselines ',
 'xonsh /  xonsh ',
 'owid /  covid-19-data ']

In [14]:
repos = [repo_ready.replace('\c','') for repo_ready in repo_ready]
repos

['yuchenlin /  rebiber ',
 'qutebrowser /  qutebrowser ',
 'sdushantha /  wifi-password ',
 'PaddlePaddle /  PaddleOCR ',
 'microsoft /  qlib ',
 'mportesi /  sorting_algo_visualizer ',
 'mobvoi /  wenet ',
 'celery /  celery ',
 'PyTorchLightning /  pytorch-lightning ',
 'donnemartin /  data-science-ipython-notebooks ',
 'public-apis /  public-apis ',
 'miloyip /  game-programmer ',
 'facebook /  prophet ',
 'jina-ai /  jina ',
 'streamlit /  streamlit ',
 'apache /  airflow ',
 'JaidedAI /  EasyOCR ',
 'allenai /  allennlp ',
 'aws /  aws-cli ',
 'geerlingguy /  ansible-for-devops ',
 'spulec /  moto ',
 'spack /  spack ',
 'openai /  baselines ',
 'xonsh /  xonsh ',
 'owid /  covid-19-data ']

#### Display all the image links from Walt Disney wikipedia page

In [15]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/Walt_Disney'

In [16]:
#your code
html = requests.get(url).content
soup = bs(html, 'lxml')

In [17]:
link1 = []
for i in soup.findAll('img', {'src':re.compile('.JPG')}):
    link1.append(i.get('src'))
    
link1

['//upload.wikimedia.org/wikipedia/commons/thumb/d/df/Walt_Disney_1946.JPG/220px-Walt_Disney_1946.JPG',
 '//upload.wikimedia.org/wikipedia/commons/thumb/1/1a/Walt_Disney_Grave.JPG/170px-Walt_Disney_Grave.JPG',
 '//upload.wikimedia.org/wikipedia/commons/thumb/a/a9/Disney_Display_Case.JPG/170px-Disney_Display_Case.JPG']

In [18]:
link2 = []
for j in soup.findAll('img', {'src':re.compile('.jpg')}):
    link2.append(j.get('src'))
    
link2

['//upload.wikimedia.org/wikipedia/commons/thumb/c/c4/Walt_Disney_envelope_ca._1921.jpg/220px-Walt_Disney_envelope_ca._1921.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Newman_Laugh-O-Gram_%281921%29.webm/220px-seek%3D2-Newman_Laugh-O-Gram_%281921%29.webm.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Trolley_Troubles_poster.jpg/170px-Trolley_Troubles_poster.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/7/71/Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_of_Review_Magazine.jpg/170px-Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_of_Review_Magazine.jpg',
 '//upload.wikimedia.org/wikipedia/en/thumb/4/4e/Steamboat-willie.jpg/170px-Steamboat-willie.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/5/57/Walt_Disney_1935.jpg/170px-Walt_Disney_1935.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/c/cd/Walt_Disney_Snow_white_1937_trailer_screenshot_%2813%29.jpg/220px-Walt_Disney_Snow_white_1937_

In [19]:
link3 = []
link4 = []
for a in link1:
    link3.append('https://' + a)

for b in link2:
    link4.append('https://' + b)
    
link3+link4

['https:////upload.wikimedia.org/wikipedia/commons/thumb/d/df/Walt_Disney_1946.JPG/220px-Walt_Disney_1946.JPG',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/1/1a/Walt_Disney_Grave.JPG/170px-Walt_Disney_Grave.JPG',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/a/a9/Disney_Display_Case.JPG/170px-Disney_Display_Case.JPG',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/c/c4/Walt_Disney_envelope_ca._1921.jpg/220px-Walt_Disney_envelope_ca._1921.jpg',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Newman_Laugh-O-Gram_%281921%29.webm/220px-seek%3D2-Newman_Laugh-O-Gram_%281921%29.webm.jpg',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Trolley_Troubles_poster.jpg/170px-Trolley_Troubles_poster.jpg',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/7/71/Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_of_Review_Magazine.jpg/170px-Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_

#### Retrieve an arbitary Wikipedia page of "Python" and create a list of links on that page

In [20]:
# This is the url you will scrape in this exercise
url ='https://en.wikipedia.org/wiki/Python' 

In [21]:
#your code
html = requests.get(url).content
soup = bs(html, 'lxml')

In [22]:
lst1 = [i.text for i in soup.findAll('a')]
lst1

['',
 'Jump to navigation',
 'Jump to search',
 'Python',
 'python',
 'Pythons',
 'Python (genus)',
 '1 Computing',
 '2 People',
 '3 Roller coasters',
 '4 Vehicles',
 '5 Weaponry',
 '6 Other uses',
 '7 See also',
 'edit',
 'Python (programming language)',
 'CMU Common Lisp',
 'PERQ 3',
 'edit',
 'Python of Aenus',
 'Python (painter)',
 'Python of Byzantium',
 'Python of Catana',
 'Python Anghelo',
 'edit',
 'Python (Efteling)',
 'Python (Busch Gardens Tampa Bay)',
 'Python (Coney Island, Cincinnati, Ohio)',
 'edit',
 'Python (automobile maker)',
 'Python (Ford prototype)',
 'edit',
 'Python (missile)',
 'Python (nuclear primary)',
 'Colt Python',
 'edit',
 'PYTHON',
 'Python (film)',
 'Python (mythology)',
 'Monty Python',
 'Python (Monty) Pictures',
 'edit',
 'Cython',
 'Pyton',
 'Pithon',
 '',
 'disambiguation',
 'internal link',
 'https://en.wikipedia.org/w/index.php?title=Python&oldid=997582414',
 'Categories',
 'Disambiguation pages',
 'Human name disambiguation pages',
 'Disambig

In [23]:
lst2 = []

for j in lst1:
    if j != '':
        lst2.append(j)
        
lst2

['Jump to navigation',
 'Jump to search',
 'Python',
 'python',
 'Pythons',
 'Python (genus)',
 '1 Computing',
 '2 People',
 '3 Roller coasters',
 '4 Vehicles',
 '5 Weaponry',
 '6 Other uses',
 '7 See also',
 'edit',
 'Python (programming language)',
 'CMU Common Lisp',
 'PERQ 3',
 'edit',
 'Python of Aenus',
 'Python (painter)',
 'Python of Byzantium',
 'Python of Catana',
 'Python Anghelo',
 'edit',
 'Python (Efteling)',
 'Python (Busch Gardens Tampa Bay)',
 'Python (Coney Island, Cincinnati, Ohio)',
 'edit',
 'Python (automobile maker)',
 'Python (Ford prototype)',
 'edit',
 'Python (missile)',
 'Python (nuclear primary)',
 'Colt Python',
 'edit',
 'PYTHON',
 'Python (film)',
 'Python (mythology)',
 'Monty Python',
 'Python (Monty) Pictures',
 'edit',
 'Cython',
 'Pyton',
 'Pithon',
 'disambiguation',
 'internal link',
 'https://en.wikipedia.org/w/index.php?title=Python&oldid=997582414',
 'Categories',
 'Disambiguation pages',
 'Human name disambiguation pages',
 'Disambiguation pag

#### Number of Titles that have changed in the United States Code since its last release point 

In [24]:
# This is the url you will scrape in this exercise
url = 'http://uscode.house.gov/download/download.shtml'

In [25]:
#your code
html = requests.get(url).content
soup = bs(html, 'lxml')

titles = soup.findAll('div', class_=['usctitle', 'usctitlechanged'])[2::]
titles

[<div class="usctitle" id="us/usc/t1">
 
           Title 1 - General Provisions <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitle" id="us/usc/t2">
 
           Title 2 - The Congress
 
         </div>,
 <div class="usctitle" id="us/usc/t3">
 
           Title 3 - The President <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitle" id="us/usc/t4">
 
           Title 4 - Flag and Seal, Seat of Government, and the States <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitle" id="us/usc/t5">
 
           Title 5 - Government Organization and Employees <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitlechanged" id="us/usc/t6">
 
           Title 6 - Domestic Security
 
         </div>,
 <div class="usctitle" id="us/usc/t7">
 
           Title 7 - Agriculture
 
         </div>,
 <div class="usctitle" id="us/usc/t8">
 
           Title 

In [26]:
len(titles)

54

#### A Python list with the top ten FBI's Most Wanted names 

In [27]:
# This is the url you will scrape in this exercise
url = 'https://www.fbi.gov/wanted/topten'

In [28]:
#your code 
html = requests.get(url).content
soup = bs(html, 'lxml')

fbi = [i.text for i in soup.findAll('h3', class_='title')]
fbi

['\nALEJANDRO ROSALES CASTILLO\n',
 '\nARNOLDO JIMENEZ\n',
 '\nJASON DEREK BROWN\n',
 '\nALEXIS FLORES\n',
 '\nJOSE RODOLFO VILLARREAL-HERNANDEZ\n',
 '\nEUGENE PALMER\n',
 '\nRAFAEL CARO-QUINTERO\n',
 '\nROBERT WILLIAM FISHER\n',
 '\nBHADRESHKUMAR CHETANBHAI PATEL\n',
 '\nYASER ABDEL SAID\n']

In [29]:
fbi1 = []

for i in fbi:
    fbi1.append(i.split('\n')[1])
    
fbi1

['ALEJANDRO ROSALES CASTILLO',
 'ARNOLDO JIMENEZ',
 'JASON DEREK BROWN',
 'ALEXIS FLORES',
 'JOSE RODOLFO VILLARREAL-HERNANDEZ',
 'EUGENE PALMER',
 'RAFAEL CARO-QUINTERO',
 'ROBERT WILLIAM FISHER',
 'BHADRESHKUMAR CHETANBHAI PATEL',
 'YASER ABDEL SAID']

####  20 latest earthquakes info (date, time, latitude, longitude and region name) by the EMSC as a pandas dataframe

In [30]:
# This is the url you will scrape in this exercise
url = 'https://www.emsc-csem.org/Earthquake/'

In [31]:
#your code
html = requests.get(url).content
soup = bs(html, 'lxml')

In [32]:
info_date = soup.findAll('td', {'class':'tabev6'})[:20]
info_lat = soup.findAll('td', {'class':'tabev1'})[:40]
info_reg = soup.findAll('td', {'class':'tb_region'})[:20]

In [33]:
lat_long = [i.text.strip().split('\n') for i in info_lat]

lat = lat_long[0::2]
long = lat_long[1::2]

print(lat, long)

[['37.20'], ['8.16'], ['19.25'], ['19.23'], ['37.20'], ['24.13'], ['5.13'], ['37.84'], ['38.92'], ['37.88'], ['8.13'], ['31.35'], ['7.97'], ['17.92'], ['37.90'], ['34.83'], ['38.91'], ['8.19'], ['19.16'], ['61.70']] [['3.74'], ['82.89'], ['155.82'], ['155.82'], ['3.70'], ['67.16'], ['103.75'], ['26.70'], ['122.88'], ['26.85'], ['82.94'], ['68.61'], ['82.94'], ['66.86'], ['26.91'], ['25.06'], ['122.88'], ['82.90'], ['155.48'], ['55.23']]


In [43]:
region = [j.get_text().strip().split('\n') for j in info_reg]
region

[['SPAIN'],
 ['PANAMA-COSTA RICA BORDER REGION'],
 ['ISLAND OF HAWAII, HAWAII'],
 ['ISLAND OF HAWAII, HAWAII'],
 ['SPAIN'],
 ['SALTA, ARGENTINA'],
 ['SOUTHERN SUMATRA, INDONESIA'],
 ['DODECANESE ISLANDS, GREECE'],
 ['NORTHERN CALIFORNIA'],
 ['DODECANESE ISLANDS, GREECE'],
 ['PANAMA-COSTA RICA BORDER REGION'],
 ['SAN JUAN, ARGENTINA'],
 ['SOUTH OF PANAMA'],
 ['PUERTO RICO REGION'],
 ['DODECANESE ISLANDS, GREECE'],
 ['CRETE, GREECE'],
 ['NORTHERN CALIFORNIA'],
 ['PANAMA-COSTA RICA BORDER REGION'],
 ['ISLAND OF HAWAII, HAWAII'],
 ['SOUTH SHETLAND ISLANDS']]

In [36]:
datetime = [k.text.strip().split('\n') for k in info_date]

datetime

[['earthquake2021-01-29\xa0\xa0\xa016:50:25.116min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa016:44:07.023min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa016:33:43.133min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa016:30:11.937min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa016:03:14.81hr 04min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa015:46:09.01hr 21min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa015:36:19.01hr 31min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa015:34:51.91hr 32min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa015:33:42.61hr 33min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa015:23:50.11hr 43min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa015:09:30.01hr 57min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa015:05:13.02hr 02min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa015:03:30.02hr 03min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa015:03:17.22hr 04min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa014:36:41.02hr 30min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa014:33:06.72hr 34min ago'],
 ['earthquake2021-01-29\

In [44]:
date = []
time = []

for a in range(len(datetime)):
    path_date = '[0-9]+-[0-9]+-[0-9]+'
    path_time = '[0-9]+:[0-9]+:[0-9]+'
    date.extend([re.findall(path_date, a) for a in datetime[a]])
    time.extend([re.findall(path_time, a) for a in datetime[a]])
    
columns = [date, time, lat, long, region]
col = [[a for [a] in b] for b in columns]

In [46]:
earthquake = pd.DataFrame(list(zip(col[0], col[1], col[2], col[3], col[4])),
                         columns = ['date','time','lat','long', 'region'])

In [47]:
earthquake

date      time    lat    long                           region
0   2021-01-29  16:50:25  37.20    3.74                            SPAIN
1   2021-01-29  16:44:07   8.16   82.89  PANAMA-COSTA RICA BORDER REGION
2   2021-01-29  16:33:43  19.25  155.82         ISLAND OF HAWAII, HAWAII
3   2021-01-29  16:30:11  19.23  155.82         ISLAND OF HAWAII, HAWAII
4   2021-01-29  16:03:14  37.20    3.70                            SPAIN
5   2021-01-29  15:46:09  24.13   67.16                 SALTA, ARGENTINA
6   2021-01-29  15:36:19   5.13  103.75      SOUTHERN SUMATRA, INDONESIA
7   2021-01-29  15:34:51  37.84   26.70       DODECANESE ISLANDS, GREECE
8   2021-01-29  15:33:42  38.92  122.88              NORTHERN CALIFORNIA
9   2021-01-29  15:23:50  37.88   26.85       DODECANESE ISLANDS, GREECE
10  2021-01-29  15:09:30   8.13   82.94  PANAMA-COSTA RICA BORDER REGION
11  2021-01-29  15:05:13  31.35   68.61              SAN JUAN, ARGENTINA
12  2021-01-29  15:03:30   7.97   82.94                  SOUTH OF PANAMA
13  2021-01-29  15:03:17  17.92   66.86               PUERTO RICO REGION
14  2021-01-29  14:36:41  37.90   26.91       DODECANESE ISLANDS, GREECE
15  2021-01-29  14:33:06  34.83   25.06                    CRETE, GREECE
16  2021-01-29  14:28:29  38.91  122.88              NORTHERN CALIFORNIA
17  2021-01-29  14:20:00   8.19   82.90  PANAMA-COSTA RICA BORDER REGION
18  2021-01-29  14:18:48  19.16  155.48         ISLAND OF HAWAII, HAWAII
19  2021-01-29  14:10:39  61.70   55.23           SOUTH SHETLAND ISLANDS

#### Display the date, days, title, city, country of next 25 hackathon events as a Pandas dataframe table

In [ ]:
# This is the url you will scrape in this exercise
url ='https://hackevents.co/hackathons'

In [ ]:
#your code

#### Count number of tweets by a given Twitter account.

You will need to include a ***try/except block*** for account names not found. 
<br>***Hint:*** the program should count the number of tweets for any provided account

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
#your code

#### Number of followers of a given twitter account

You will need to include a ***try/except block*** in case account/s name not found. 
<br>***Hint:*** the program should count the followers for any provided account

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
#your code

#### List all language names and number of related articles in the order they appear in wikipedia.org

In [48]:
# This is the url you will scrape in this exercise
url = 'https://www.wikipedia.org/'

In [81]:
#your code
html = requests.get(url).content
soup = bs(html, 'lxml')

In [82]:
wiki = [i.text for i in soup.findAll('div', class_='central-featured-lang')]
wiki

['\n\nEnglish\n6\xa0235\xa0000+ articles\n\n',
 '\n\n日本語\n1\xa0250\xa0000+ 記事\n\n',
 '\n\nDeutsch\n2\xa0528\xa0000+ Artikel\n\n',
 '\n\nEspañol\n1\xa0655\xa0000+ artículos\n\n',
 '\n\nРусский\n1\xa0693\xa0000+ статей\n\n',
 '\n\nFrançais\n2\xa0292\xa0000+ articles\n\n',
 '\n\nItaliano\n1\xa0669\xa0000+ voci\n\n',
 '\n\n中文\n1\xa0172\xa0000+ 條目\n\n',
 '\n\nPolski\n1\xa0452\xa0000+ haseł\n\n',
 '\n\nPortuguês\n1\xa0053\xa0000+ artigos\n\n']

In [83]:
wiki1 = [i.replace('\n','') for i in wiki]
wiki1

['English6\xa0235\xa0000+ articles',
 '日本語1\xa0250\xa0000+ 記事',
 'Deutsch2\xa0528\xa0000+ Artikel',
 'Español1\xa0655\xa0000+ artículos',
 'Русский1\xa0693\xa0000+ статей',
 'Français2\xa0292\xa0000+ articles',
 'Italiano1\xa0669\xa0000+ voci',
 '中文1\xa0172\xa0000+ 條目',
 'Polski1\xa0452\xa0000+ haseł',
 'Português1\xa0053\xa0000+ artigos']

In [84]:
wiki2 = [i.split('\xa0') for i in wiki1]
wiki2

[['English6', '235', '000+ articles'],
 ['日本語1', '250', '000+ 記事'],
 ['Deutsch2', '528', '000+ Artikel'],
 ['Español1', '655', '000+ artículos'],
 ['Русский1', '693', '000+ статей'],
 ['Français2', '292', '000+ articles'],
 ['Italiano1', '669', '000+ voci'],
 ['中文1', '172', '000+ 條目'],
 ['Polski1', '452', '000+ haseł'],
 ['Português1', '053', '000+ artigos']]

In [85]:
wiki3 = [' '.join(i) for i in wiki2]
wiki3

['English6 235 000+ articles',
 '日本語1 250 000+ 記事',
 'Deutsch2 528 000+ Artikel',
 'Español1 655 000+ artículos',
 'Русский1 693 000+ статей',
 'Français2 292 000+ articles',
 'Italiano1 669 000+ voci',
 '中文1 172 000+ 條目',
 'Polski1 452 000+ haseł',
 'Português1 053 000+ artigos']

#### A list with the different kind of datasets available in data.gov.uk 

In [ ]:
# This is the url you will scrape in this exercise
url = 'https://data.gov.uk/'

In [ ]:
#your code 

#### Top 10 languages by number of native speakers stored in a Pandas Dataframe

In [ ]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/List_of_languages_by_number_of_native_speakers'

In [ ]:
#your code

### BONUS QUESTIONS

#### Scrape a certain number of tweets of a given Twitter account.

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
# your code

#### IMDB's Top 250 data (movie name, Initial release, director name and stars) as a pandas dataframe

In [ ]:
# This is the url you will scrape in this exercise 
url = 'https://www.imdb.com/chart/top'

In [ ]:
# your code

#### Movie name, year and a brief summary of the top 10 random movies (IMDB) as a pandas dataframe.

In [ ]:
#This is the url you will scrape in this exercise
url = 'http://www.imdb.com/chart/top'

In [ ]:
#your code

#### Find the live weather report (temperature, wind speed, description and weather) of a given city.

In [ ]:
#https://openweathermap.org/current
city = city=input('Enter the city:')
url = 'http://api.openweathermap.org/data/2.5/weather?'+'q='+city+'&APPID=b35975e18dc93725acb092f7272cc6b8&units=metric'

In [ ]:
# your code

#### Book name,price and stock availability as a pandas dataframe.

In [ ]:
# This is the url you will scrape in this exercise. 
# It is a fictional bookstore created to be scraped. 
url = 'http://books.toscrape.com/'

In [ ]:
#your code